Enable `AbsentOrWrongFileLicense` detekt rule

dector opened the pull request:
- Enable new built-in `AbsentOrWrongFileLicense`.
  - Generate baseline (will create follow-up) issue to get rid of it.
  - Baseline is also required because `LongMethod` now is reacting on constructors as well.
  - Upgrade detekt to ~~`1.7.4`~~ `1.9.1`.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] ~~**Tests**: This PR includes thorough tests or an explanation of why it does not~~
- [x] ~~**Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not~~
- [x] ~~**Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.~~

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

liuche commented on buildSrc/src/main/java/Dependencies.kt:
```diff
@@ -65,79 +65,109 @@ object Deps {
 
     const val mozilla_concept_engine = "org.mozilla.components:concept-engine:${Versions.mozilla_android_components}"
     const val mozilla_concept_push = "org.mozilla.components:concept-push:${Versions.mozilla_android_components}"
-    const val mozilla_concept_tabstray = "org.mozilla.components:concept-tabstray:${Versions.mozilla_android_components}"
```
Can you  revert the formatting changes to all these lines? That way our git blame won't pick up the wrong changes.

dector replied:
Will do

dector commented:
- Rebased with master
  - Removed formatting changes
  - Updated detekt to 1.9.1 (no changes required for this codebase)

liuche commented:
Hi @dector! I'm seeing a few problems while trying to push this PR:
* `./gradlew assembleGeckoBetaDebug` is failing with your changes
* There are 2 detekt problems that probably need to be added to the baseline (`UseCases.kt` and `SessionControlController.kt`)

Thanks for your patience! This is really close and I want to help you get this landed :)

dector commented:
@liuche 

About (2): Fixed. Rebased with master again and updated baseline.

About (1): `assembleGeckoBetaDebug` is failing because new `detekt` versions uses Kotlin `1.3.72`. We have two options:

    a) Raise project Kotlin version to `1.3.72` (as it'll be used with Gradle by default) and use `@OptIn()` for certain classes. I can create issue and fix it there (or in this PR).
    b) Add `ExperimentalCoroutinesApi` to `-Xuse-experimental`. This will silent all experimental coroutines API warnings.

I can do both. But which option is more preferable?

liuche commented:
Hi @dector! Could you do option b)? We are getting close to shipping a big release, so we don't want to add instability of updating the kotlin version right now.

Thank you!

dector commented:
Sure. Will do.

dector commented:
What was done:

1) Rebased with `master`.
2) Applied `ExperimentalCoroutinesApi` to project `:app`. Please, check it in a2f185d. The issue here that we can't apply it to root project because `detekt` task has no dependency on coroutines artifact.

3) Removed `-Xskip-runtime-version-check` as it's [not valid anymore](https://discuss.kotlinlang.org/t/about-skip-runtime-version-check-related-warnings/4949).

4) I made some research and it seems that because of the way Gradle [resolves dependencies](https://docs.gradle.org/current/userguide/dependency_resolution.html#sec:version-conflict) Kotlin Stdlib 1.3.72 is used in project de-facto. Here are `--scan` results:

<details><summary>gekoBetaDebugCompileClasspath</summary>
<img src="https://user-images.githubusercontent.com/265872/85219503-5e411f00-b3b5-11ea-8f3a-5ec4c752684c.png" />
</details>

<details><summary>gekoBetaDebugRuntimeClasspath</summary>
<img src="https://user-images.githubusercontent.com/265872/85219504-600ae280-b3b5-11ea-9853-73cb0885cdbd.png" />
</details>

liuche commented:
I'm so sorry @dector - but every time I push your PR, it fails the pre-push hook. Could you run that as part of your push? https://github.com/mozilla-mobile/fenix/#pre-push-hooks

dector commented:
Sorry, I was using [old hook version](https://github.com/mozilla-mobile/fenix/blob/8af55652be5c7c3058e5f53ffaa2925d67bd8602/config/pre-push-recommended.sh). ^_^

See latest commit.

liuche commented on app/build.gradle:
```diff
@@ -753,3 +753,10 @@ if (gradle.hasProperty('localProperties.autoPublish.application-services.dir'))
+
+tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
+    kotlinOptions.freeCompilerArgs += [
+            "-Xuse-experimental=kotlinx.coroutines.ExperimentalCoroutinesApi",
```
@NotWoods I went through the app:dependencies while investigating this PR and saw that we end up bumping the kotlin version to 1.3.72 implicitly. Curious if bumping the coroutines version could help resolve some of these annotation problems?

NotWoods replied:
We do this in AC as well (albeit on a per-component basis), as many Android Components use `ExperimentalCoroutinesApi`. However, the newer Kotlin version does make this much easier to avoid. 

The current annotation forces any caller to use it, and their callers, and their callers, winding all the way up the call stack.

Newer Kotlin versions (1.7 IIRC) let you just write `OptIn(ExperimentalCoroutinesApi)` and will just annotate that call, without putting requirements on its callers.

Since we've cut a release branch we can update the Kotlin version then pull those changes in here.

dector replied:
@liuche Sure!

> Is it because you'd need to annotate the detekt config? Could you help me understand why this is necessary, or what other options there are? (Is the alternative to annotate each usage?)

I guess applying `-Xuse` was agreed in [this thread](https://github.com/mozilla-mobile/fenix/pull/9705#issuecomment-637889887).

Alternatively we can annotate each usage with `@UseExperimental` to avoid compiler warnings that breaks our build (`allWarningsAsErrors=true`).

liuche replied:
Ah, you're totally right. I thought at the time that it would be simpler to not bump the Kotlin version, but when looking at other dependencies, it seems like we already use 1.3.72, so now that I've dug a little more into it, option a) seems to be better, to switch to @OptIn or @UseExperimental.

Sorry for all the back and forth.

dector replied:
Not a problem! Will rebase this PR with @NotWoods changes.

dector commented:
Rebased, squashed, updated.

liuche commented:
Merged in #12748 !!!!
